### PR TITLE
Compatibility: SQL_SELECT_LIMIT worked in show statement;

### LINKED ISSUE #33407: SQL_SELECT_LIMIT worked with show columns in MYSQL but not in TIDB
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE t1(a INTEGER, b INTEGER, c INTEGER, d INTEGER);
INSERT INTO t1 VALUES(1,2,3,4);
INSERT INTO t1 VALUES(4,3,1,2);

SET @@sql_select_limit=1;
PREPARE stmt FROM "SHOW COLUMNS FROM t1";
EXECUTE stmt;
PREPARE stmt FROM "select * FROM t1";
EXECUTE stmt;
```
### 2. What did you expect to see? (Required)
```sql

mysql> CREATE TABLE t1(a INTEGER, b INTEGER, c INTEGER, d INTEGER);
Query OK, 0 rows affected (0.01 sec)

mysql> INSERT INTO t1 VALUES(1,2,3,4);
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO t1 VALUES(4,3,1,2);
Query OK, 1 row affected (0.00 sec)

mysql> 
mysql> SET @@sql_select_limit=1;
Query OK, 0 rows affected (0.00 sec)

mysql> PREPARE stmt FROM "SHOW COLUMNS FROM t1";
Query OK, 0 rows affected (0.00 sec)
Statement prepared

mysql> EXECUTE stmt;
+-------+------+------+-----+---------+-------+
| Field | Type | Null | Key | Default | Extra |
+-------+------+------+-----+---------+-------+
| a     | int  | YES  |     | NULL    |       |
+-------+------+------+-----+---------+-------+
1 row in set (0.01 sec)

mysql> PREPARE stmt FROM "select * FROM t1";
Query OK, 0 rows affected (0.00 sec)
Statement prepared

mysql> EXECUTE stmt;
+------+------+------+------+
| a    | b    | c    | d    |
+------+------+------+------+
|    1 |    2 |    3 |    4 |
+------+------+------+------+
1 row in set (0.00 sec)
```
### 3. What did you see instead (Required)
```sql

mysql> drop table t1;
Query OK, 0 rows affected (0.33 sec)

mysql> CREATE TABLE t1(a INTEGER, b INTEGER, c INTEGER, d INTEGER);
Query OK, 0 rows affected (0.12 sec)

mysql> INSERT INTO t1 VALUES(1,2,3,4);
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO t1 VALUES(4,3,1,2);
Query OK, 1 row affected (0.01 sec)

mysql> 
mysql> SET @@sql_select_limit=1;
Query OK, 0 rows affected (0.00 sec)

mysql> PREPARE stmt FROM "SHOW COLUMNS FROM t1";
Query OK, 0 rows affected (0.00 sec)

mysql> EXECUTE stmt;
+-------+---------+------+------+---------+-------+
| Field | Type    | Null | Key  | Default | Extra |
+-------+---------+------+------+---------+-------+
| a     | int(11) | YES  |      | NULL    |       |
| b     | int(11) | YES  |      | NULL    |       |
| c     | int(11) | YES  |      | NULL    |       |
| d     | int(11) | YES  |      | NULL    |       |
+-------+---------+------+------+---------+-------+
4 rows in set (0.00 sec)

mysql> PREPARE stmt FROM "select * FROM t1";
Query OK, 0 rows affected (0.00 sec)

mysql> EXECUTE stmt;
+------+------+------+------+
| a    | b    | c    | d    |
+------+------+------+------+
|    1 |    2 |    3 |    4 |
+------+------+------+------+
1 row in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```sql
 Release Version: v5.4.0
Edition: Community
Git Commit Hash: 55f3b24c1c9f506bd652ef1d162283541e428872
Git Branch: heads/refs/tags/v5.4.0
UTC Build Time: 2022-01-25 08:41:05
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by Benjamin2037**
issue pingcap/tidb#33407
### What problem does this PR solve?

Issue Number: close #33407

Problem Summary:

### What is changed and how it works?
The supported Show stmts are go through to find out what stmts should be influenced by SelectLimit value on return result set row numbers.  The Limit executor is added into plan tree for show stmts to archieve the requirement.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: executor/show.go [hunk 1]
```diff
@@ -128,113 +129,121 @@ func (e *ShowExec) Next(ctx context.Context, req *chunk.Chunk) error {
 }
 
 func (e *ShowExec) fetchAll(ctx context.Context) error {
+	var err error = nil
+
+	// Temporary disable select limit to avoid miss the result.
+	oldSelectLimit := e.ctx.GetSessionVars().SelectLimit
+	e.ctx.GetSessionVars().SelectLimit = math.MaxUint64
```
Comment (lcwangchao) [inline executor/show.go:136]: Maybe using defer to recover oldSelectLimit is better.

Another question is I'm not very understand why we should set `SelectLimit` to math.MaxUint64

Comment (Benjamin2037) [reply executor/show.go:136]: Yes, Have changed restore SelectLimit logic with defer.

The math.MaxUint64 is used as default of SelectLimit and it means no limit on return result set row numbers.

File: executor/show.go [hunk 2]
```diff
@@ -128,113 +129,122 @@ func (e *ShowExec) Next(ctx context.Context, req *chunk.Chunk) error {
 }
 
 func (e *ShowExec) fetchAll(ctx context.Context) error {
+	var err error = nil
+
+	// Temporary disable select limit to avoid miss the result.
```
Comment (lcwangchao) [inline executor/show.go:134]: Please add some comments to describe why we need to set `SelectLimit` to Max

File: executor/show.go [hunk 3]
```diff
@@ -128,113 +129,122 @@ func (e *ShowExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
+	oldSelectLimit := e.ctx.GetSessionVars().SelectLimit
+	e.ctx.GetSessionVars().SelectLimit = math.MaxUint64
```
Comment (lcwangchao) [inline executor/show.go:136]: Please add some comments to describe why we need to set SelectLimit to Max

**Review by lcwangchao — COMMENTED** [review]
Rest LGTM

**Comment by Benjamin2037** [comment]
> Rest LGTM

Please Approve

**Review by lcwangchao — APPROVED** [review]

File: executor/show.go [hunk 4]
```diff
@@ -139,6 +140,24 @@ func (e *ShowExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
+	// cause a filter effect on result set that may exclude the qualified record outside of
+	// result set.
+	// Finally, when above result set, may not include qualified record, is returned to up
+	// level show stmt's selection, which really applies the filter operation on returned
+	// result set, it may return empty result to client.
+	oldSelectLimit := e.ctx.GetSessionVars().SelectLimit
+	e.ctx.GetSessionVars().SelectLimit = math.MaxUint64
+	defer func() {
+		// Restore session Var SelectLimit value.
+		e.ctx.GetSessionVars().SelectLimit = oldSelectLimit
```
Comment (xhebox) [inline executor/show.go:158]: Can we test this?

**Review by xhebox — COMMENTED** [review]
Rest LGTM

**Review by xhebox — APPROVED** [review]

**Pull request merged**